man/machinery.1.md
# Machinery — A Systems Management Toolkit for Linux

## SYNOPSIS

`machinery` SUBCOMMAND \[options\] <br>
`machinery` help [SUBCOMMAND]


## DESCRIPTION

Machinery is a systems management toolkit for Linux. It supports configuration
discovery, system validation, and service migration. Machinery's core concept
is the complete representation of a system by a universal system description.
System descriptions are managed independently of the described systems which
allows for system state conservation and offline preparation of modifications.

Machinery has a set of commands which work with this system description.
These commands can be combined to form work flows. Machinery is targeted at the
system administrator of the data center.

More information about `machinery` and its usage can be found at
http://machinery-project.org or in the documentation which can be opened using
`machinery man --html`.


NEWS
# Machinery Release Notes

* Fix treatment of the rpm `(replaced)` flag (gh#SUSE/machinery#1897)
* Fix repository recognition on older Debian systems
* Inspection of unmanaged-files is now also using the faster machinery-helper
  when a remote user is used
* Show all available information for unmanaged files in HTML view (gh#SUSE/machinery#1905)
* Move the machinery documentation from man page to HTML format
* Implement "--extract-metadata" option for the unmanaged-files inspection to
  retrieve the file meta (e.g. owner, group, size) data without extracting the
  files.
* Fixed counting of files in unmanaged directories (empty directories now report zero files
  as one would expect)

## Version 1.17.0 - Mon Jan 25 16:49:58 CET 2016 - thardeck@suse.de

* Fix reset button in HTML view (gh#SUSE/machinery#1899)
* Align config-files scope in HTML view (gh#SUSE/machinery#1814)
* Allow closing file dialog (in WebUI) by pressing ESC
* Support inspection of Ubuntu 14.04 systems
* New `--ssh-identity-file` option to specify a SSH private key when inspecting
  a remote system
* New `--ssh-port` option to specify an SSH port when inspecting a remote system

## Version 1.16.4 - Thu Jan 14 18:35:47 CET 2016 - thardeck@suse.de

* The unmanaged-files scope filter will not include "/boot" by default anymore.
  In order to have the same behavior as before run:
  `machinery inspect --skip-files=/boot`
* Fix crash during image build on SLES12 SP1 (gh#SUSE/machinery#1828)

## Version 1.16.3 - Thu Jan 07 17:57:27 CET 2016 - thardeck@suse.de

* Remove concept of "supported systems" in favor of letting Machinery run
  everywhere. Whether a system can be built, exported, etc will be validated
  from its related task by checking if the propper tooling is installed
  (Fixes bnc#959558)
* Remove hint for `show`, when `inspect-container` is run with `--show` option
* Rephrase the startup messages and warn messages when starting the HTTP
  server for compare, show, or serve system descriptions
* Change `serve` command to always list all system descriptions

## Version 1.16.2 - Tue Nov 24 13:17:18 CET 2015 - thardeck@suse.de

* Fix crash when the user does not have enough permissions to create the image
  directory for the `build` command (gh#SUSE/machinery#1667)
* Fix wrong hint for docker containers while running `analyze` command
  (gh#SUSE/machinery#1632)
* Add a new hint when using the `--public` option for the `serve` command that it
  makes all descriptions publicly available
* Fix Machinery failing to inspect changed config files in case of restrictive
  permissions (gh#SUSE/machinery#1609)
* Remove hint for `show`, when `inspect` is run with `--show` option
  (gh#SUSE/machinery#1648)
* Fix repository inspection on hosts that have a LANGUAGE variable set
* Empty scopes now display a message in HTML & CLI views (gh#SUSE/machinery#1615)
* Fix SLES 11 SP4 os inspector output to be similar to SLES 11 SP3
* Fix Autoyast export on openSUSE Leap and Tumbleweed (gh#SUSE/machinery#1627)
* Allow building of openSUSE Leap and Tumbleweed on openSUSE Leap and Tumbleweed
* Allow building of SLES11 and SLES12 on openSUSE Leap
* Add HTML landing page, which lists all system descriptions on your system

## Version 1.16.1 - Mon Nov 09 17:55:15 CET 2015 - thardeck@suse.de

* Fix type recognition in case of empty files for config- and changed-managed-files
  inspectors (gh#SUSE/machinery#1612)

## Version 1.16.0 - Fri Nov 06 14:20:47 CET 2015 - thardeck@suse.de

* The error message "There is already a server with the same port running" has
  been cleaned up and is no longer misleading (gh#SUSE/machinery#1541)
* Add OpenSUSE Leap as supported system
* HTML comparison is now available without the experimental feature flag
* Add error message to recommend using `--name` when a container with a slash is
  inspected (gh#SUSE/machinery#1490)
* Fixed crash during inspection of a folder which contains special characters
  in its name (gh#SUSE/machinery#1580)
* Speed up inspection of config and changed managed files
* Fix docker inspection when specifying a docker image with tag
  (gh#SUSE/machinery#1491)
* GConf XML files are no longer treated as binary files in HTML view
  (gh#SUSE/machinery#1154)
* Add current outgoing ip and export directory to Autoyast README.md during
  export (gh#SUSE/machinery#1543)
* Handle invalid utf-8 characters in /etc/passwd
* Fix: "changed" link in compare view is visible while scope is collapsed but
  doesn't do anything (gh#SUSE/machinery#1555)

## Version 1.15.0 - Mon Oct 26 17:42:25 CET 2015 - thardeck@suse.de

* No information is cut off in the HTML view
* Files containing special characters can now be downloaded
  (gh#SUSE/machinery#1497)
* Links to common elements work in HTML when all scopes are collapsed
  (gh#SUSE/machinery#1469)
* Show if description was result of a container inspection
  (gh#SUSE/machinery#1495)
* Disable nav bar buttons when scope has been excluded (gh#SUSE/machinery#1199)
* Show details of elements with changed attributes for all scopes in
  comparison view
* Show progress indicator while loading file content in HTML view

## Version 1.14.2 - Wed Oct 14 18:00:51 CEST 2015 - thardeck@suse.de

* Fix typo in HTML package view  (gh#SUSE/machinery#1493)

## Version 1.14.1 - Fri Oct 09 18:30:13 CEST 2015 - thardeck@suse.de

* Fix: Clean up binding the server for HTML view to IP addresses
  (gh#SUSE/machinery#1341)
* Fix: Make links to sections with common elements clearer in HTML comparison
  view (gh#SUSE/machinery#1222)
* Fix: XML files are no longer treated as binary files in HTML view
  (gh#SUSE/machinery#1154)
* Fix: Scrolling issue of file view (gh#SUSE/machinery#1155)

## Version 1.14.0 - Tue Oct 06 21:26:12 CEST 2015 - thardeck@suse.de

* The machinery-helper is now built during gem-installation on x86_64
  machines if Go is available. The RPM package is still shipped with
  a precompiled version.
* Allow limiting the `list` command output to certain system descriptions

lib/machinery.rb
# Copyright (c) 2013-2016 SUSE LLC
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 3 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact SUSE LLC.
#
# To contact SUSE about this file by physical or electronic mail,
# you may find current contact information at www.suse.com

require "ostruct"
require "json"
require "abstract_method"
require "cheetah"
require "tmpdir"
require "tempfile"
require "time"
require "logger"
require "erb"
require "yaml"
require "uri"
require "gli"
require "json-schema"
require "haml"
require "kramdown"
require "find"
require "pathname"
require "nokogiri"
require "socket"
require "diffy"
require "sinatra/base"
require "mimemagic"

require_relative "try"
require_relative "machinery_logger"
require_relative "zypper"
require_relative "rpm"
require_relative "array"
require_relative "object"
require_relative "constants"
require_relative "manifest"
require_relative "system_description"
require_relative "version"
require_relative "tarball"
require_relative "exceptions"
require_relative "inspector"
require_relative "system"
require_relative "local_system"
require_relative "remote_system"
require_relative "current_user"
require_relative "inspect_task"
require_relative "inspector"
require_relative "build_task"
require_relative "exporter"
require_relative "kiwi_config"
require_relative "renderer"
require_relative "show_task"
require_relative "compare_task"
require_relative "remove_task"
require_relative "list_task"
require_relative "system_description_store"
require_relative "logged_cheetah"
require_relative "renderer_helper"
require_relative "export_task"
require_relative "helper"
require_relative "deploy_task"
require_relative "analyze_config_file_diffs_task"
require_relative "copy_task"
require_relative "scope"
require_relative "ui"
require_relative "validate_task"
require_relative "migration"
require_relative "upgrade_format_task"
require_relative "hint"
require_relative "mountpoints"
require_relative "config_base"
require_relative "config"
require_relative "config_task"
require_relative "autoyast"
require_relative "system_description_memory_store"
require_relative "scope_file_store"
require_relative "json_validator"
require_relative "json_validation_error_cleaner"
require_relative "file_validator"
require_relative "element_filter"
require_relative "filter"
require_relative "filter_option_parser"
require_relative "file_scope"
require_relative "system_file"
require_relative "scope_file_access_flat"
require_relative "scope_file_access_archive"
require_relative "man_task"
require_relative "comparison"
require_relative "serve_html_task"
require_relative "file_diff"
require_relative "server"
require_relative "html"
require_relative "machinery_helper"
require_relative "workload_mapper"
require_relative "containerize_task"
require_relative "workload_mapper_dsl"
require_relative "containerized_app"
require_relative "move_task"
require_relative "docker_system"
require_relative "managed_files_database"
require_relative "rpm_database"
require_relative "dpkg_database"
require_relative "tee_io"

Dir[File.join(Machinery::ROOT, "plugins", "**", "*.rb")].each { |f| require(f) }

# this file needs be loaded last, because it immediately collects the loaded inspectors
require_relative "cli"
